### DISCUSSION TITLE
Alternatives to BinaryCompatOnlyAttribute

### DISCUSSION THREAD

**Discussion by ds5678** [dotnet/csharplang#7717] [Language Ideas]
# Context

Since the main motivation for [`BinaryCompatOnlyAttribute`](https://github.com/dotnet/csharplang/pull/7707) is to resolve overload resolution problems, I propose this be resolved by changing `ObsoleteAttribute` instead.

# Idea 1: Make Obsolescence Affect Overload Resolution

We define an "obsolete context" where a containing type or member is obsolete. In this context, overload resolution is treated the same as it is today.

Outside of an obsolete context, obsolete members are treated as a last resort for resolution, perhaps even after extension members with the same signature.

# Idea 2: Add a Boolean Property to ObsoleteAttribute

```cs
partial class ObsoleteAttribute
{
    bool BinaryCompatOnly { get; set; }
}
```

**Comment by 333fred** [comment]
These have already been suggested on the original issue, we don't need a separate discussion to fracture the conversation and make it harder to follow.

**Reply by ds5678** [reply]
Sorry, I searched for other discussions and didn't find any.

**Reply by ds5678** [reply]
I found the issue: #7706.
